### PR TITLE
EKF: mag_control: Reset yaw to mag when restarting fusion

### PR DESCRIPTION
**Pull request by mhkabir**
When mag fusion was stopped (e.g when EV yaw data fusion was started) and restarted, the yaw angle was not reset to that provided by the mag, leading to filter divergence.

The fix is to make sure we reset the yaw when re-starting mag fusion.

**Before this PR:**
image

**After this PR:**
image

**EV Data is stopped at the 79s mark, and restarted at the 95s mark**

Also see https://github.com/PX4/Firmware/issues/14840 where this discovered.

Flight log which illustrates old issue: https://logs.px4.io/plot_app?log=c0a43438-fb28-45fd-9a5b-92329748fb93. 

The fix was tested by running replay on the above log.

### CODE REVIEW

File: EKF/control.cpp
Comment (kamilritz) [inline EKF/control.cpp:396]:
```suggestion
	} else if ((_control_status.flags.ev_pos || _control_status.flags.ev_vel || _control_status.flags.ev_yaw)
		   && isTimedOut(_time_last_ext_vision, (uint64_t)_params.reset_timeout_max)) {
if(_control_status.flags.ev_yaw) {_mag_yaw_reset_req=true};
```

**Review by kamilritz — CHANGES_REQUESTED** [review]
Thanks for solving this. 
At the moment we have a variable `_mag_yaw_reset_req` that is used to request mag resets. I suggest that we use this variable for solving this. 
I understand that it is not nice that each heading sensor needs to activate the default heading sensor, but I prefer to have not different solutions for this.

**Comment by mhkabir** [comment]
Can you explain your rationale?

**Comment by kamilritz** [comment]
Sure. Apparently there are certain situations where a heading reset is needed. At the moment for this to happen the part of the software that realizes that a reset is necessary, requests a reset through the `_mag_yaw_reset_req` variable. 

You are adding a second option for the reset to happen by `restarting_fusion ||`. This is changing the functionality of the `run*YawReset` function a lot as it is now also responsible for finding out if it should do a reset. Previously it was only told from other parts of the software. The function will do now both which I find not that clean.

I would prefer if the nature of the function stays the same and you command it to do a reset from else where by setting the `_mag_yaw_reset_req` variable. As a benefit you don't have to add the line `const bool restarting_fusion = !_control_status.flags.mag_hdg && canRunMagFusion();` twice.

If you don't want to trigger the reset from the `controlExternalVision()` call, you could also do this from the controlMagFusion function with `_mag_yaw_reset_req = !_control_status.flags.mag_hdg;` right after [this line](https://github.com/PX4/ecl/blob/master/EKF/mag_control.cpp#L73).

If you want to do me a favor you could also rename the `canRunMagFusion()` to `shouldRunMagFusion()`, which would make it more clear.

**Comment by mhkabir** [comment]
@kamilritz your changes are applied, and I rebased. Will give this state a test now.

**Comment by kamilritz** [comment]
@mhkabir Did you have a chance to test the new version?

**Comment by mhkabir** [comment]
Ah yes, it worked fine. I forgot to pull logs, but I flew it several times and things seemed to be fine.

**Comment by mhkabir** [comment]
I have updated the change_indicator reference outptuts @kamilritz @priseborough. This should be good to merge, and is a crucial bugfix for PX4 v1.11.

File: EKF/mag_control.cpp
Comment (kamilritz) [inline EKF/mag_control.cpp:74]: I guess this is changing the change indicator output. I am a bit surprised that this change is also affecting the output for a model that is not using any vision measurement.

**Review by kamilritz — DISMISSED** [review]
It is a bit a surprise that the change indicator need to be updated. But the difference is minor.

**Comment by mhkabir** [comment]
@kamilritz @bresch can we get this in

**Comment by kamilritz** [comment]
@mhkabir This should be ready to get in. I do not have merge rights, so there is nothing I can do about it.

File: EKF/mag_control.cpp
Comment (bresch) [reply EKF/mag_control.cpp:78]: This will continuously request resets in `mag_3D` mode.

**Review by bresch — CHANGES_REQUESTED** [review]

**Review by priseborough — DISMISSED** [review]
OK with this going in if the change indicator file is updated

**Comment by dagar** [comment]
What happened here? Should we rebase this?

**Comment by priseborough** [comment]
Yes it needs to be rebased and the change indicator file updated which was what was holding up the approval and merge last time.

**Comment by dagar** [comment]
Rebased.

**Pull request closed without merging**
